src/libNode/MicroBlockProcessing.cpp

```diff
@@ -726,62 +934,144 @@ bool Node::CheckMicroBlockTimestamp()
     return true;
 }
 
-bool Node::CheckLegitimacyOfTxnHashes(vector<unsigned char>& errorMsg)
+bool Node::ProcessTransactionWhenShardBackup(const vector<TxnHash>& tranHashes,
+                                             vector<TxnHash>& missingtranHashes)
 {
-    lock(m_mutexReceivedTransactions, m_mutexSubmittedTransactions);
-    lock_guard<mutex> g(m_mutexReceivedTransactions, adopt_lock);
-    lock_guard<mutex> g2(m_mutexSubmittedTransactions, adopt_lock);
+    LOG_MARKER();
+
+    auto findFromCreated = [this](Transaction& t, const TxnHash& th) -> bool {
+        lock_guard<mutex> g(m_mutexCreatedTransactions);
+
+        auto& hashIdx = m_createdTransactions.get<MULTI_INDEX_KEY::TXN_ID>();
+        if (!hashIdx.size())
+        {
+            return false;
+        }
+
+        // auto it = find_if(
+        //     begin(m_createdTransactions), end(m_createdTransactions),
+        //     [&th](const Transaction& t) { return t.GetTranID() == th; });
+
+        // if (m_createdTransactions.end() == it)
+        // {
+        //     LOG_GENERAL(WARNING, "txn is not found");
+        //     return false;
+        // }
 
-    auto const& receivedTransactions
-        = m_receivedTransactions[m_mediator.m_currentEpochNum];
-    auto const& submittedTransactions
-        = m_submittedTransactions[m_mediator.m_currentEpochNum];
+        // t = move(*it);
+        // m_createdTransactions.erase(it);
+
+        auto it = hashIdx.find(th);
+
+        if (hashIdx.end() == it)
+        {
+            LOG_GENERAL(WARNING, "txn is not found");
+            return false;
+        }
+
+        t = move(*it);
+        hashIdx.erase(it);
+
+        return true;
+    };
+
+    std::list<Transaction> curTxns;
+
+    for (const auto& tranHash : tranHashes)
+    {
+        Transaction t;
+
+        if (findFromCreated(t, tranHash))
+        {
+            curTxns.emplace_back(t);
+        }
+        else
+        {
+            missingtranHashes.emplace_back(tranHash);
+        }
+    }
+
+    if (!missingtranHashes.empty())
+    {
+        return true;
+    }
+
+    if (!VerifyTxnsOrdering(curTxns))
+    {
+        return false;
+    }
+
+    uint64_t blockNum = m_mediator.m_currentEpochNum;
+
+    auto appendOne = [this, &blockNum](const Transaction& t) {
+        lock_guard<mutex> g(m_mutexProcessedTransactions);
+        auto& processedTransactions = m_processedTransactions[blockNum];
```

Comment:
blockNum is needless and here we can just use m_processedTransactions[m_mediator.m_currentEpochNum]